### PR TITLE
`EntityTag` full compatibility

### PR DESCRIPTION
**Pull request by tal5**
Finishes up `EntityTag` to be closer to the server-side implementation, this includes: support for un-spawned entities, properties, and fake entities (ones created by the server).

## Implementation details

- **Un-spawned entities:** an entity instance is created from the type so that it can be adjusted and used as usual. This is not perfect, but should work for now to allow receiving, using, and sending back a generic `EntityTag`.
- **Fake entities:** an `isFake` boolean is stored, which is used in all the relevant places (e.g. having `e@fake:` instead of `e@` in `identify` methods).
- **Properties** are implemented pretty much as usual - added an `EntityProperty` class for them to implement, and a base/example one.

## Additons

- `Utilities#uuidFromString` - a simple util to avoid having several `try/catch` blocks all over `valueOf`.
- `Utilities#echoErrorByContext` - to remove the boilerplate `context == null || context.showErrors()` checks before every error in `valueOf` methods.
- `EntityTag#getEntityForID` - gets an entity by it's UUID.
- `EntityTag#isSpawned` - pretty obvious lol, checks if the entity is currently spawned.
- `EntityTag#is(EntityType)` - to check if an entity is of a specific type, mainly for `describes` in properties.

### CODE REVIEW

File: src/main/java/com/denizenscript/clientizen/objects/EntityTag.java [hunk 1]
```diff
@@ -1,48 +1,108 @@
...
+            Utilities.echoErrorByContext(context, "valueOf EntityTag returning null: invalid entity type '" + string + "'.");
+            return null;
         }
-        return null;
+        return entityByType;
+    }
+
+    public static Entity getEntityForID(UUID uuid) {
+        return MinecraftClient.getInstance().world.getEntityLookup().get(uuid);
     }
```
[the remark this answers is not in the document]
Comment (tal5) [reply src/main/java/com/denizenscript/clientizen/objects/EntityTag.java]: Fair enough, that's probably a clearer name

File: src/main/java/com/denizenscript/clientizen/objects/EntityTag.java [hunk 2]
```diff
@@ -54,15 +114,25 @@ public static boolean matches(String string) {
...
         return entity;
     }
 
+    public boolean isSpawned() {
+        return uuid != null && getEntity() != null && entity.isAlive();
+    }
+
+    public boolean is(EntityType<?> type) {
+        return getEntity().getType() == type;
+    }
```
[the remark this answers is not in the document]
Comment (tal5) [reply src/main/java/com/denizenscript/clientizen/objects/EntityTag.java:131]: Was thinking about how it's method calls would look - it'd basically always be `is(EntityType.X)`, making `isType(EntityType.X)` feel like a redundant `Type`

File: src/main/java/com/denizenscript/clientizen/objects/properties/entity/EntityProperty.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+        return object.getEntity();
+    }
+
+    public <T extends Entity> T as(EntityType<T> type) {
+        return type.downcast(getEntity());
+    }
+
+    public boolean is(EntityType<?> type) {
+        return getEntity().getType() == type;
+    }
```
[the remark this answers is not in the document]
Comment (tal5) [reply src/main/java/com/denizenscript/clientizen/objects/properties/entity/EntityProperty.java:21]: https://github.com/acikek/Clientizen/pull/20#discussion_r1215697616

File: src/main/java/com/denizenscript/clientizen/objects/EntityTag.java [hunk 4]
```diff
@@ -81,27 +151,50 @@ public ObjectTag getObjectAttribute(Attribute attribute) {
...
+            Text displayName = entity.getCustomName();
+            if (displayName != null) {
+                debuggable += "<LG>/<GR>" + displayName.getString();
             }
-            debuggable += ")";
+            return debuggable + ')';
         }
-        return debuggable;
+        return "<LG>e@<Y>" + entity.getType().getUntranslatedName() + PropertyParser.getPropertiesDebuggable(this);
     }
```
[the remark this answers is not in the document]
Comment (tal5) [reply src/main/java/com/denizenscript/clientizen/objects/EntityTag.java]: Good idea, done

File: src/main/java/com/denizenscript/clientizen/objects/EntityTag.java [hunk 5]
```diff
@@ -1,48 +1,108 @@
...
+            Entity entity = getEntityForID(uuid);
+            if (entity != null) {
+                EntityType<?> entityType = EntityType.get(string).orElse(null);
+                // If the value isn't a valid entity type then just let it through, as we can't verify entity scripts
+                if (entityType != null && entity.getType() != entityType) {
+                    Utilities.echoErrorByContext(context, "valueOf EntityTag returning null: UUID '" + uuidString + "' is valid, but doesn't match the provided entity type.");
+                    return null;
+                }
+                return new EntityTag(entity);
+            }
```
[the remark this answers is not in the document]
Comment (tal5) [reply src/main/java/com/denizenscript/clientizen/objects/EntityTag.java]: Added a `valueOfByType` method for better logic flow

**Review by acikek — APPROVED** [review]

**Pull request merged**